### DISCUSSION TITLE
After extended playing, font glyphs unload causing unreadable text.

### DISCUSSION THREAD

**Discussion by zachmanthethird** [ppy/osu#15924] [Q&A]
Essentially, after about an hour of gameplay, any font that I haven't used starts unloading(?) glyphs. My best guess is memory overloading or misusage (I have 16GB of RAM), but it could just be a rendering issue. The first time I encountered it was by pressing Ctrl-F11 and even numbers weren't displaying. The second time was when certain beatmap metadata (Japanese characters specifically) refused to load leaving awkward spaces. This time, I pressed Ctrl-F7 to semi-labelled options; I finally had the thought process to collect data to present.

Once I encountered the error, I took as many screenshots to illustrate the problem (7 +2 with Ctrl+F3 enabled). I then restarted the client and observed the same points under as similar circumstances as possible to show the comparison (7 +1 with Ctrl+F3 enabled). Additionally, I've zipped 3 related log files (from before restarting) which seem normal but may reveal more to you.

Preview screenshot with error:
![osu_2021-12-03_19-14-19](https://user-images.githubusercontent.com/74262407/144696237-44cd33ac-a7f6-45fd-8150-cae61f7b6028.png)

Preview screenshot after restart:
![osu_2021-12-03_19-20-00](https://user-images.githubusercontent.com/74262407/144696242-7376e010-a4ec-4a1d-af19-40bc48f24240.png)

Lazer Versions 2021.1120.0 - 2021.1202.0
Full .zip (GitHub does not like me uploading this apparently, but I don't want to spam this with files):
https://drive.google.com/file/d/1vg6QxYo3AZKaTnMLotsgx50lKm5liaLM

**Comment by peppy** [comment]
What's your GPU? Is it always the same characters?

**Reply by zachmanthethird** [reply]
I'm not at my computer, but I'm using an i3-10100 integrated graphics. I'll get you more details once I get back.
The characters are not the same. The fonts aren't even necessarily the same. I distinctly remember ぱぴぽ？ not loading correctly once, and it's perfectly fine in both screenshots here. Sometimes it's letters, sometimes it's numbers, there doesn't seem to be any consistency. The only pattern I can see is that the missing characters only appear on interfaces I hadn't yet used since starting client.

**Reply by zachmanthethird** [reply]
Yeah that's definitely what it was (referring to ぱぴぽ？). I have two screenshots, but they're not super clear because I didn't move the Ctrl+F3 screen away.
From Windows About:
`Processor	Intel(R) Core(TM) i3-10100 CPU @ 3.60GHz   3.60 GHz`
`Installed RAM	16.0 GB (15.8 GB usable)`
I'm using Intel(R) UHD Graphics 630 with driver version 27.20.100.8681
Let me know where to pull more information if necessary :D.

**Reply by peppy** [reply]
You mention viewing the ctrl+f3 screen - do you have any screenshots of this? It might provide some more information that can help us track down the source of the issue.

**Reply by smoogipoo** [reply]
@peppy there's screenshots in the zip in the OP.

Some glyphs are definitely looking to disappear:
![image](https://user-images.githubusercontent.com/1329837/144785976-f90fc88f-393b-4950-b0ea-7a3a32ab703b.png)
(See 131.)

But I have no clue what could be causing this at this point. Just to confirm it's completely fine after a restart? That should rule out the on-disk cache...

**Reply by peppy** [reply]
There's no way that can happen post load, as far as I'm concerned. Must be failing at the point of uploading the new (missing) glyph to the texture atlas.

**Reply by zachmanthethird** [reply]
Yes. Full visibility is completely restored on restarting the client. I'll see if I can't get it to happen again to provide more information.

**Reply by smoogipoo** [reply]
> There's no way that can happen post load, as far as I'm concerned

The zip files contain screenshots of song select having missing glyphs, so I'm finding this hard to believe given how little beatmaps they have loaded.
Can you confirm whether those glyphs were previously visible in the same screens prior to them disappearing @zachmanthethird ?

**Reply by peppy** [reply]
@smoogipoo Would would you envisage this happening from a framework/gl level? Glyphs are uploaded on a per glyph basis at new coordinates. Previous glyphs are not touched. It sounds unfeasible for arbitrary glyphs in the texture to be erased perfectly within their bounds.

**Reply by smoogipoo** [reply]
I'd agree that it sounds infeasible. We don't have any way to remove glyphs from an atlas.

**Reply by zachmanthethird** [reply]
Just happened again, this time in the Ctrl+F2 screen. I'll play a bit more to see if something happens, then reload client and try to load as many interfaces as possible. That way I'll see if I can confirm if they were previously visible.
https://user-images.githubusercontent.com/74262407/144793375-de2fe6f7-2824-4708-b61f-f135fd839647.mp4

**Reply by peppy** [reply]
Just to confirm: what we're interested in confirming is that the missing characters were never visible before the bug occurs (ie. they are visible from the first time they are "displayed", or not displayed in this case).

**Reply by zachmanthethird** [reply]
Okay, I've loaded every interface I can think of... I'll let you know if anything disappears.
![osu_2021-12-05_22-05-09](https://user-images.githubusercontent.com/74262407/144795977-43c8718f-56d2-405b-babc-39c9562192cd.png)
![osu_2021-12-05_22-05-18](https://user-images.githubusercontent.com/74262407/144795984-e5e57c72-fdfd-42b9-bf0e-a1adc8936096.png)

**Reply by zachmanthethird** [reply]
> Can you confirm whether those glyphs were previously visible in the same screens prior to them disappearing @zachmanthethird ?

It appears as though once loaded correctly, they do not unload. Checking all screens on launch, none of the characters have disappeared since. I now have 19 full character atlases with no gaps. This leaves my best assumption to be that the glyphs are loading incorrectly into the atlas in the first place.
![osu_2021-12-05_23-38-38](https://user-images.githubusercontent.com/74262407/144806284-8915cf43-123b-412a-8034-ec37fbfd4ce5.png)
I'll continue to test and see if I can't determine the point when glyphs start to load incorrectly.

**Reply by peppy** [reply]
Thanks for the confirmation. It does sound like some uploads are failing and we're not handling errors in said cases.

**Comment by zachmanthethird** [comment]
I while ago I stopped noticing this bug and believed it to have fixed itself. However, when creating a bracket.json yesterday, it showed itself consistently, and I was finally able to get a recording with the Textures menu open:

https://user-images.githubusercontent.com/74262407/149617436-620ab8c8-0cc8-4ed3-823b-25387495ecdd.mp4

I uploaded the whole clip to show that I wasn't doing anything other than opening the window, but I'd like to direct your attention to 1:02 
where the majority of the atlases are loaded in.
[bracket.json.txt](https://github.com/ppy/osu/files/7874693/bracket.json.txt)
[network.log.txt](https://github.com/ppy/osu/files/7874697/network.log.txt)
[performance.log.txt](https://github.com/ppy/osu/files/7874698/performance.log.txt)
[runtime.log.txt](https://github.com/ppy/osu/files/7874699/runtime.log.txt)

**Reply by peppy** [reply]
Thanks, about what I expected. Next is going to be figuring out how that can happen. Likely a GL error we're missing.

**Reply by bdach** [reply]
for what it's worth i attempted to reproduce this using the provided `bracket.json` but to no avail, i don't see any missing glyphs.

**Reply by peppy** [reply]
It's 100% a gpu/driver issue, so won't reproduce under any circumstances unless you match the setup.

**Comment by smoogipoo** [comment]
Can you try again on the latest release?

**Reply by zachmanthethird** [reply]
After [a suggestion from bdach](https://github.com/ppy/osu/discussions/15472#discussioncomment-2077494), I've disabled hardware acceleration. Still happens.
Version 2022.128.0
![image](https://user-images.githubusercontent.com/74262407/151919261-442ce5e2-5c10-4e9f-b535-1c920f6f8262.png)
[performance.log.txt](https://github.com/ppy/osu/files/7976198/performance.log.txt) | [network.log.txt](https://github.com/ppy/osu/files/7976200/network.log.txt) | [bracket.json.txt](https://github.com/ppy/osu/files/7976201/bracket.json.txt)
